### DISCUSSION TITLE
Multipart POST support

### DISCUSSION THREAD

**Discussion by upsetbit** [vlang/v#11844] [Ideas]
Thank you for all the work invested in V.

I was trying to make a multipart post request and noticed that the standard library does not have support for such. I've messed a little bit on `net.http` and made this: https://github.com/caiertl/v/commit/165cd390b92d0804b5c65c7cd09536045f43e5af

It worked fine for standard ASCII files (text files in general), but anything "binary" (.jpg, .png, .mp3 etc) doesn't. The program don't panic, it just hangs indefinitely.

Is there any plans to add this feature? Am I missing something on the code I wrote to make it work?

**Comment by mcastorina** [comment]
PRs are welcome :)

I gave your code a quick review:

```v
+ uheaders.join('\r\n')
+ 'Connection: close\r\n\r\n'
```
The `Connection` header will not be on a new line and is corrupting the last added header.

---

```v
r_data = req.file.map(fn (b byte) string { return b.ascii_str() }).join('')
```
This might be why it works for text only and not binary data.

**Comment by upsetbit** [comment]
@mcastorina thank you for taking the time to review.

I've enabled verbose to see how my changes affected the payload, this is what get:

* POST w/ `multipart/form-data`
https://user-images.githubusercontent.com/12755789/113500777-5ac7f000-9510-11eb-9e27-349c7e4ab961.png

* POST w/ `application/json`
https://user-images.githubusercontent.com/12755789/113500791-78955500-9510-11eb-8d01-f016caeeaf71.png

---

> The Connection header will not be on a new line and is corrupting the last added header.

Are you saying that the `Connection` header must have an empty line above it? Like this:

```
POST /repos/foo/bar/releases/00000000/assets?name=org.png HTTP/1.1
Host: uploads.github.com
User-Agent: v.http
Content-Length: 17313
Accept: application/vnd.github.v3+json
Authorization: token lorem-ipsum-sit-dolor-amet
Content-Type: multipart/form-data

Connection: close

*raw binary data here*
```

This scenario raised a Bad Request (400) from the server.

---

> This might be why it works for text only and not binary data.

I've initially used `read_file` from `os` to get the file content:

```v
foo := os.read_file('path/to/foo') or { panic('could not read foo') }
req := Request{ url: '', file: foo }
```

`read_file` returns a string optional. `read_bytes` appears to be a better choice because I can handle each individual byte. This is why `file` at the `Request` struct is defined as `[]byte`.

Escaping the bytes (with `str_escaped`) or any other "transformation" corrupts the file. This is where I'm stuck...

I could simply do

```v
r_data = req.file.bytestr()
```

But that doesn't help either. I was reading HTTP/1.1 RFCs about multipart and there is a vagueness about "raw data". To my understanding, it means a bytes array. Looking further into the code, at `http_do`, I saw this:

```v
s := req.build_request_headers(method, host_name, path)
mut client := net.dial_tcp(host) ?
// TODO this really needs to be exposed somehow
client.write(s.bytes()) ?
```

So the string generated at `build_request_headers` is being transformed to bytes before the TCP write. Maybe find a way to construct the request without using any string?

**Comment by nedimf** [comment]
I would suggest to look at net/smtp. I think smtp module now supports sending of raw files.

**Comment by mcastorina** [comment]
> Are you saying that the Connection header must have an empty line above it?

No, the headers look correct in that image. I thought the code was incorrectly concatenating them like so:
```v
arr := ['hello', 'world']
println(arr.join('\n\r') + 'foo')
// hello
// worldfoo
```

But somehow it's working correctly, so ignore that comment :)

---

> Maybe find a way to construct the request without using any string?

I think this would be the way to go. Technically V strings do not care about null termination but imo `[]byte` makes a lot more sense for the body than a string.

**Comment by mcastorina** [comment]
For now, you can convert `[]byte` to `string` with the `.bytestr()` method:

```v
x := [byte(`A`) 0 `C`]
s := x.bytestr()
println(s.bytes()) // [A, `\0`, C]
```

**Comment by Terisback** [comment]
I've done a encoder before @caiertl 
Is that what you were talking about?
https://github.com/Terisback/discord.v/blob/master/rest/formdata/formdata.v
